File: extension/src/test/unit/documentScope.test.ts
import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import { isManagedDocument } from '../../core/documentScope';

const base = {
    fsPath: '/w/service.architecture.toml',
    relativePath: 'service.architecture.toml',
    patterns: ['**/*.architecture.toml'] as string[],
    activatedUris: new Set<string>(),
};

test('.c4d documents are always managed, globs or no globs', () => {
    assert.equal(isManagedDocument({ ...base, uri: 'file:///w/model.c4d', fsPath: '/w/model.c4d', relativePath: 'model.c4d', patterns: [] }), true);
    assert.equal(isManagedDocument({ ...base, uri: 'file:///w/model.c4d', fsPath: '/w/model.c4d', relativePath: 'model.c4d' }), true);
});

test('.toml managed when a configured glob matches', () => {
    assert.equal(isManagedDocument({ ...base, uri: 'file:///w/service.architecture.toml' }), true);
    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'file:///w/arch/service.toml',
            fsPath: '/w/arch/service.toml',
            relativePath: 'arch/service.toml',
            patterns: ['arch/*.toml'],
        }),
        true,
    );
});

test('.toml NOT managed for plain TOML files', () => {
    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'file:///w/plain.toml',
            fsPath: '/w/plain.toml',
            relativePath: 'plain.toml',
        }),
        false,
    );
});

test('.toml NOT managed when no patterns configured (opt-in only)', () => {
    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'file:///w/service.architecture.toml',
            patterns: [],
        }),
        false,
    );
});

test('explicit activation manages any TOML regardless of globs', () => {
    const activated = new Set(['file:///w/plain.toml']);

    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'file:///w/plain.toml',
            fsPath: '/w/plain.toml',
            relativePath: 'plain.toml',
            patterns: [],
            activatedUris: activated,
        }),
        true,
    );
});

test('activation is keyed on the URI, not the file type', () => {
    const activated = new Set(['untitled:Untitled-1']);

    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'untitled:Untitled-1',
            fsPath: 'Untitled-1',
            relativePath: undefined,
            patterns: [],
            activatedUris: activated,
        }),
        true,
    );
});

test('other file types are never managed', () => {
    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'file:///w/notes.txt',
            fsPath: '/w/notes.txt',
            relativePath: 'notes.txt',
        }),
        false,
    );
});

test('extension match is case-insensitive (uppercase .TOML opts in via its glob)', () => {
    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'file:///w/MODEL.TOML',
            fsPath: '/w/MODEL.TOML',
            relativePath: 'MODEL.TOML',
            patterns: ['**/*.TOML'],
        }),
        true,
    );
    assert.equal(
        isManagedDocument({
            ...base,
            uri: 'file:///w/model.C4D',
            fsPath: '/w/model.C4D',
            relativePath: 'model.C4D',
            patterns: [],
        }),
        true,
    );
});

File: idea-plugin/src/test/kotlin/dev/djarvur/c4drill/core/DocumentScopeTest.kt
package dev.djarvur.c4drill.core

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Scoping rule tests: .c4d always managed, .toml opt-in only, everything else untouched. */
class DocumentScopeTest {
    @Test
    fun `c4d files are always managed`() {
        assertTrue(
            isManagedDocument(
                ScopeDecisionInput(fsPath = "/work/models/main.c4d"),
            ),
        )
        assertTrue(
            isManagedDocument(
                ScopeDecisionInput(fsPath = "/work/other.C4D"),
            ),
        )
    }

    @Test
    fun `plain toml is untouched without patterns`() {
        assertFalse(isManagedDocument(ScopeDecisionInput(fsPath = "/work/go.mod.toml")))
        assertFalse(
            isManagedDocument(
                ScopeDecisionInput(fsPath = "/work/Cargo.toml", patterns = emptyList()),
            ),
        )
    }

    @Test
    fun `toml with a matching relative pattern is managed`() {
        assertTrue(
            isManagedDocument(
                ScopeDecisionInput(
                    fsPath = "/work/models/main.toml",
                    relativePath = "models/main.toml",
                    patterns = listOf("models/*.toml"),
                ),
            ),
        )
    }

    @Test
    fun `toml with an absolute pattern is managed`() {
        assertTrue(
            isManagedDocument(
                ScopeDecisionInput(
                    fsPath = "/work/models/main.toml",
                    patterns = listOf("/work/models/*.toml"),
                ),
            ),
        )
    }

    @Test
    fun `non-matching toml stays untouched`() {
        assertFalse(
            isManagedDocument(
                ScopeDecisionInput(
                    fsPath = "/work/src/config.toml",
                    relativePath = "src/config.toml",
                    patterns = listOf("models/*.toml"),
                ),
            ),
        )
    }

    @Test
    fun `explicit activation wins over everything`() {
        val activated = setOf("/work/Cargo.toml")

        assertTrue(isManagedDocument(ScopeDecisionInput(fsPath = "/work/Cargo.toml", activatedPaths = activated)))
        // ...but only for the activated path.
        assertFalse(isManagedDocument(ScopeDecisionInput(fsPath = "/work/other.toml", activatedPaths = activated)))
    }

    @Test
    fun `double star patterns match nested models`() {
        assertTrue(
            isManagedDocument(
                ScopeDecisionInput(
                    fsPath = "/work/arch/systems/main.toml",
                    relativePath = "arch/systems/main.toml",
                    patterns = listOf("**/*.toml"),
                ),
            ),
        )
    }

    @Test
    fun `only toml and c4d extensions are considered`() {
        assertFalse(isManagedDocument(ScopeDecisionInput(fsPath = "/work/models/main.yaml", patterns = listOf("**/*"))))
        assertFalse(isManagedDocument(ScopeDecisionInput(fsPath = "/work/models/c4dtoml")))
    }
}
